adapter/internal/discovery/xds/rate_limiter_cache.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	if _, ok := r.apiLevelRateLimitPolicies[org][vHost]; !ok {
+		r.apiLevelRateLimitPolicies[org][vHost] = make(map[string][]*rls_config.RateLimitDescriptor)
+	}
+
+	r.apiLevelRateLimitPolicies[org][vHost][apiID] = rlsConfigs
+	return nil
+}
+
+// DeleteAPILevelRateLimitPolicies deletes inline Rate Limit policies added with the API.
+func (r *rateLimitPolicyCache) DeleteAPILevelRateLimitPolicies(org, vHost, apiID string) {
+	delete(r.apiLevelRateLimitPolicies[org][vHost], apiID)
+}
+
+func getRateLimitPolicy(policies map[string]*mgw.APIRateLimitPolicy, policyName string) (*rls_config.RateLimitPolicy, error) {
+	policy, ok := policies[policyName]
+	if !ok {
+		return nil, fmt.Errorf("rate limit policy %q not defined", policyName)
+	}
+
+	unit, err := getRateLimitUnit(policy.SpanUnit)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rls_config.RateLimitPolicy{
+		Unit:            unit,
+		RequestsPerUnit: policy.Count,
+	}, nil
+}
+
+func (r *rateLimitPolicyCache) generateRateLimitConfig(label string) *rls_config.RateLimitConfig {
+	var orgDescriptors []*rls_config.RateLimitDescriptor
+
+	// Generate API level rate limit configurations
+	for org, orgPolicies := range r.apiLevelRateLimitPolicies {
+		var vHostDescriptors []*rls_config.RateLimitDescriptor
+		for vHost, vHostPolicies := range orgPolicies {
+			var apiDescriptors []*rls_config.RateLimitDescriptor
+			for apiID, apiPolicies := range vHostPolicies {
+				// Configure API Level rate limit policies only if, the API is deployed to the gateway label
+				// Check API deployed to the gateway label
+				if arrayContains(orgIDOpenAPIEnvoyMap[org][apiID], label) {
+					apiDescriptors = append(apiDescriptors, apiPolicies...)
+				}
+			}
+			vHostDescriptor := &rls_config.RateLimitDescriptor{
+				Key:         "vhost",
+				Value:       vHost,
+				Descriptors: apiDescriptors,
+			}
+			vHostDescriptors = append(vHostDescriptors, vHostDescriptor)
+		}
+		orgDescriptor := &rls_config.RateLimitDescriptor{
+			Key:         "org",
+			Value:       org,
+			Descriptors: vHostDescriptors,
+		}
+		orgDescriptors = append(orgDescriptors, orgDescriptor)
+	}
+
+	return &rls_config.RateLimitConfig{
+		Name:        "Default",
+		Domain:      "Default",
+		Descriptors: orgDescriptors,
+	}
+}
+
+func (r *rateLimitPolicyCache) updateXdsCache(label string) bool {
+	rlsConf := r.generateRateLimitConfig(label)
+
+	version := fmt.Sprint(rand.Intn(maxRandomInt))
+	snap, err := gcp_cache.NewSnapshot(version, map[gcp_resource.Type][]gcp_types.Resource{
+		gcp_resource.RateLimitConfigType: {
+			rlsConf,
+		},
+	})
+	if err != nil {
+		loggers.LoggerXds.Error("Error while updating the rate limit snapshot: ", err)
+		return false
+	}
+	if err := snap.Consistent(); err != nil {
+		loggers.LoggerXds.Error("Inconsistent rate limiter snapshot: ", err)
+		return false
+	}
+
+	if err := r.xdsCache.SetSnapshot(context.Background(), label, snap); err != nil {
```

Comment:
Is this thread safe ? No need to use a mutex here ?

Reply:
It is handled in the go control plane side.
https://github.com/envoyproxy/go-control-plane/blob/b7f08b4ef20a2788772e1f88f36b591a6f897b6f/pkg/cache/v3/simple.go#L225

Btw we can do that for our internal maps. Done in https://github.com/wso2/product-microgateway/pull/3178/commits/e80a84426bc693e7105d05ffdef1b8219bbf6fa8